Just a dump of what I've looked at so far. Doing builds with the branch in #25391 (which includes this change) for convenience.

Building depends locally with `LTO=1`, i.e `make -C depends/ HOST=x86_64-apple-darwin LTO=1` works as expected, and ultimately produces working binaries, using LTO, after compiling using the depends CONFIG_SITE.

However building the same branch in Guix, i.e `time HOSTS="x86_64-apple-darwin arm64-apple-darwin" ./contrib/guix/guix-build`, bombs out during the depends build with:
```bash
/bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool -static -o libminiupnpc.a miniwget.o minixml.o igd_desc_parse.o minisoap.o miniupnpc.o upnpreplyparse.o upnpcommands.o upnperrors.o connecthostport.o portlistingparse.o receivedata.o upnpdev.o addr_is_reserved.o minissdpc.o
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: miniwget.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: minixml.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: igd_desc_parse.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: minisoap.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: miniupnpc.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: upnpreplyparse.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: upnpcommands.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: upnperrors.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: connecthostport.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: portlistingparse.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: receivedata.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: upnpdev.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: addr_is_reserved.o is not an object file (not allowed in a library)
error: /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool: file: minissdpc.o is not an object file (not allowed in a library)
make[1]: *** [Makefile:302: libminiupnpc.a] Error 1
```

Looking at `miniupnpc.o`, the object files are LLVM bitcode:
```bash
file miniupnpc/2.2.2-fcea302a1da/miniupnpc.o
miniupnpc/2.2.2-fcea302a1da/miniupnpc.o: LLVM bitcode, wrapper x86_64
```

and they should be supported if libtool in cctools is compiled with LTO support (`-DLTO_SUPPORT`), which it is:
```bash
/root/.guix-profile/bin/clang -DPACKAGE_NAME=\"cctools\" -DPACKAGE_TARNAME=\"cctools\" -DPACKAGE_VERSION=\"973.0.1\" -DPACKAGE_STRING=\"cctools\ 973.0.1\" -DPACKAGE_BUGREPORT=\"t.poechtrager@gmail.com\" -DPACKAGE_URL=\"\" -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DEMULATED_HOST_CPU_TYPE=16777223 -DEMULATED_HOST_CPU_SUBTYPE=3 -D__STDC_LIMIT_MACROS=1 -D__STDC_CONSTANT_MACROS=1 -DHAVE_EXECINFO_H=1 -DHAVE_UTIMENSAT=1 -DHAVE_BCOPY=1 -DHAVE_BCMP=1 -DHAVE_BZERO=1 -DHAVE_INDEX=1 -DHAVE_RINDEX=1 -DHAVE_STAT_ST_MTIM=1 -DHAVE_CLOCK_GETTIME=1 -I.   -I/bitcoin/depends/x86_64-apple-darwin/native/include -I../include -I../include/foreign -I../libstuff -Wall -Wno-long-long -Wno-import -Wno-format -Wno-deprecated -Wno-unused-variable -Wno-unused-private-field -Wno-unused-function -Wno-invalid-offsetof -Wno-int-conversion -Wno-char-subscripts -Wno-shift-negative-value -Wno-misleading-indentation -Wno-gnu-folding-constant -DLTO_SUPPORT -D__DARWIN_UNIX03 -D__LITTLE_ENDIAN__=1 -O3 -isystem /usr/local/include -isystem /usr/pkg/include  -DLD64_VERSION_NUM=609 -fblocks -D__arm64__ -c -o libtool.o libtool.c
```

We can further confirm that by checking that the libtool binary has symbols that would only be compiled in under LTO support, i.e [`lto_toc_nsyms`](https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L253-L258):
```bash
nm -C /bitcoin/depends/x86_64-apple-darwin/native/bin/x86_64-apple-darwin-libtool | grep lto_
0000000000437480 b lto_create
0000000000437488 b lto_create_local
0000000000437490 b lto_dispose
00000000004374c0 b lto_error
000000000040e558 T lto_free
00000000004374b8 b lto_get_cputype
000000000040e5e8 T lto_get_nlist_64
000000000040e564 T lto_get_nsyms
00000000004374a0 b lto_get_num_symbols
00000000004374a8 b lto_get_sym_attr
00000000004374b0 b lto_get_sym_name
0000000000437498 b lto_get_target
0000000000437470 b lto_handle
0000000000437478 b lto_is_object
000000000040e7a8 T lto_symbol_name
000000000040e570 T lto_toc_symbol
0000000000408230 t save_lto_member_toc_info
```
Looking at the error message, `file: miniupnpc.o is not an object file (not allowed in a library)`, there are a number of places that could be coming from inside libtool:
https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L1541
https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L1614
https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L1632
https://github.com/tpoechtrager/cctools-port/blob/6540086c5e12e9c1649fed524b527d8c1793ddc0/cctools/misc/libtool.c#L1663

--- a/depends/hosts/darwin.mk
+++ b/depends/hosts/darwin.mk
@@ -17,6 +17,7 @@ darwin_native_toolchain=native_cctools
 
 clang_prog=$(build_prefix)/bin/clang
 clangxx_prog=$(clang_prog)++
+llvm_config_prog=$(build_prefix)/bin/llvm-config
 
 clang_resource_dir=$(build_prefix)/lib/clang/$(native_clang_version)
 else
@@ -34,8 +35,10 @@ darwin_native_toolchain=
 # Source: https://lists.gnu.org/archive/html/bug-make/2017-11/msg00017.html
 clang_prog=$(shell $(SHELL) $(.SHELLFLAGS) "command -v clang")
 clangxx_prog=$(shell $(SHELL) $(.SHELLFLAGS) "command -v clang++")
+llvm_config_prog=$(shell $(SHELL) $(.SHELLFLAGS) "command -v llvm-config")
 
 clang_resource_dir=$(shell clang -print-resource-dir)
+llvm_lib_dir=$(shell $(llvm_config_prog) --libdir)
 endif
 
 cctools_TOOLS=AR RANLIB STRIP NM LIBTOOL OTOOL INSTALL_NAME_TOOL DSYMUTIL

--- a/depends/packages/native_cctools.mk
+++ b/depends/packages/native_cctools.mk
@@ -16,9 +16,17 @@ define $(package)_set_vars
   $(package)_cxx=$(clangxx_prog)
 endef
 
+ifneq ($(strip $(FORCE_USE_SYSTEM_CLANG)),)
+define $(package)_preprocess_cmds
+  mkdir -p $($(package)_staging_prefix_dir)/lib && \
+  cp $(llvm_lib_dir)/libLTO.so $($(package)_staging_prefix_dir)/lib/ && \
+  cp -f $(BASEDIR)/config.guess $(BASEDIR)/config.sub cctools
+endef
+else
 define $(package)_preprocess_cmds
   cp -f $(BASEDIR)/config.guess $(BASEDIR)/config.sub cctools
 endef
+endif
 
 define $(package)_config_cmds
   $($(package)_autoconf)

--- a/depends/packages/native_cctools.mk
+++ b/depends/packages/native_cctools.mk
@@ -7,11 +7,9 @@ $(package)_build_subdir=cctools
 $(package)_dependencies=native_libtapi
 
 define $(package)_set_vars
-  $(package)_config_opts=--target=$(host)
+  $(package)_config_opts=--target=$(host) --enable-lto-support
+  $(package)_config_opts+=--with-llvm-config=$(llvm_config_prog)
   $(package)_ldflags+=-Wl,-rpath=\\$$$$$$$$\$$$$$$$$ORIGIN/../lib
-  ifeq ($(strip $(FORCE_USE_SYSTEM_CLANG)),)
-  $(package)_config_opts+=--enable-lto-support --with-llvm-config=$(build_prefix)/bin/llvm-config
-  endif
   $(package)_cc=$(clang_prog)
   $(package)_cxx=$(clangxx_prog)
 endef

--- a/depends/packages/native_clang.mk
+++ b/depends/packages/native_clang.mk
@@ -16,10 +16,13 @@ endef
 define $(package)_stage_cmds
   mkdir -p $($(package)_staging_prefix_dir)/lib/clang/$($(package)_version)/include && \
   mkdir -p $($(package)_staging_prefix_dir)/bin && \
+  mkdir -p $($(package)_staging_prefix_dir)/include/llvm-c && \
   cp bin/clang $($(package)_staging_prefix_dir)/bin/ && \
   cp -P bin/clang++ $($(package)_staging_prefix_dir)/bin/ && \
   cp bin/dsymutil $($(package)_staging_prefix_dir)/bin/$(host)-dsymutil && \
   cp bin/llvm-config $($(package)_staging_prefix_dir)/bin/ && \
+  cp include/llvm-c/ExternC.h $($(package)_staging_prefix_dir)/include/llvm-c && \
+  cp include/llvm-c/lto.h $($(package)_staging_prefix_dir)/include/llvm-c && \
   cp lib/libLTO.so $($(package)_staging_prefix_dir)/lib/ && \
   cp -r lib/clang/$($(package)_version)/include/* $($(package)_staging_prefix_dir)/lib/clang/$($(package)_version)/include/
 endef

--- a/depends/packages/native_libtapi.mk
+++ b/depends/packages/native_libtapi.mk
@@ -13,7 +13,5 @@ define $(package)_build_cmds
 endef
 
 define $(package)_stage_cmds
-  ./install.sh && \
-  mkdir -p $($(package)_staging_prefix_dir)/include/llvm-c && \
-  cp src/llvm/include/llvm-c/lto.h $($(package)_staging_prefix_dir)/include/llvm-c
+  ./install.sh
 endef

